python/ray/data/_internal/util.py

```diff
@@ -51,3 +58,96 @@ def _check_pyarrow_version():
             )
         else:
             _VERSION_VALIDATED = True
+
+
+def _autodetect_parallelism(
+    parallelism: int,
+    cur_pg: Optional["PlacementGroup"],
+    ctx: DatasetContext,
+    reader: Optional["Reader"] = None,
+    avail_cpus: Optional[int] = None,
+) -> (int, int):
+    """Returns parallelism to use and the min safe parallelism to avoid OOMs.
+
+    This detects parallelism using on the following heuristics, applied in order:
+
+     1) We start with the default parallelism of 200.
+     2) Min block size. If the parallelism would make blocks smaller than this
+        threshold, the parallelism is reduced to avoid the overhead of tiny blocks.
+     3) Max block size. If the parallelism would make blocks larger than this
+        threshold, the parallelism is increased to avoid OOMS during processing.
+     4) Available CPUs. If the parallelism cannot make use of all the available
+        CPUs in the cluster, the parallelism is increased until it can.
+
+    Args:
+        parallelism: The user-requested parallelism, or -1 for auto-detection.
+        cur_pg: The current placement group, to be used for avail cpu calculation.
+        ctx: The current Dataset context to use for configs.
+        reader: The datasource reader, to be used for data size estimation.
+        avail_cpus: Override avail cpus detection (for testing only).
+
+    Returns:
+        Tuple of detected parallelism (only if -1 was specified), and the min safe
+        parallelism (which can be used to generate warnings about large blocks).
+    """
+    min_safe_parallelism = 1
+    max_reasonable_parallelism = 9999999
```

Comment:
I was a bit afraid this would cast the entire thing to float in some situations, so the large int seemed safer.

Reply:
Hmm `max`/`min` don't cast their outputs, so I don't think that we need to worry about that, and any future refactor that produces a float would most likely be returning infinity; such a bug should be caught by our tests.

If you're still concerned about a float sneaking in somewhere, IMO we should use `sys.maxsize` rather than this "maybe large enough" number. https://docs.python.org/3/library/sys.html#sys.maxsize

Reply:
sys.maxsize it is then